plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java

```diff
@@ -2685,6 +2694,112 @@ public Optional<ProjectionApplicationResult<ConnectorTableHandle>> applyProjecti
                 false));
     }
 
+    @Override
+    public Optional<PartialSortApplicationResult<ConnectorTableHandle>> applyPartialSort(
+            ConnectorSession session,
+            ConnectorTableHandle tableHandle,
+            Map<ColumnHandle, io.trino.spi.connector.SortOrder> columnHandleSortOrderMap)
+    {
+        if (columnHandleSortOrderMap.size() != 1) {
+            return Optional.empty();
+        }
+        IcebergTableHandle originalTableHandle = (IcebergTableHandle) tableHandle;
+        if (originalTableHandle.isSort()) {
+            return Optional.empty();
+        }
+        IcebergTableHandle icebergTableHandle = (IcebergTableHandle) tableHandle;
+        Table table;
+        table = catalog.loadTable(session, icebergTableHandle.getSchemaTableName());
+        IcebergColumnHandle sortOrderColumn = (IcebergColumnHandle) columnHandleSortOrderMap.keySet().stream().toList().get(0);
+        Set<Integer> specIdNotMatching = new HashSet<>();
+
+        // partition specs are {0:f1, 1:f1,f2, 2:f3}
+        // if query sort is `order by f2`, filter out result(specIdNotMatching) is specId=0,2. later we will use `planFile` decide whether the planed files have specId 0 or 2.
+        // if all the files spec ids does not have 0 or 2 due to some reason, then we hope the query can use our optimization.
+        table.specs().values().forEach(spec -> {
```

Comment:
if spec is no longer used, does it matter?

(can a historical spec make the feature not work?)

Reply:
It depends. The important thing is whether the plan file generated includes this historical spec. If it does, then the feature not work.